src/rust/inetstack/protocols/layer2/mod.rs

```diff
@@ -0,0 +1,149 @@
+    layer1_endpoint: Box<dyn PhysicalLayer>,
+    link_addr: MacAddress,
+}
+
+#[derive(Clone)]
+pub struct SharedLayer2Endpoint(SharedObject<Layer2Endpoint>);
+
+//======================================================================================================================
+// Associated Functions
+//======================================================================================================================
+
+impl SharedLayer2Endpoint {
+    pub fn new<P: PhysicalLayer>(config: &Config, layer1_endpoint: P) -> Result<Self, Fail> {
+        Ok(Self(SharedObject::new(Layer2Endpoint {
+            layer1_endpoint: Box::new(layer1_endpoint),
+            link_addr: config.local_link_addr()?,
+        })))
+    }
+
+    pub fn receive(&mut self) -> Result<ArrayVec<(Ethernet2Header, DemiBuffer), RECEIVE_BATCH_SIZE>, Fail> {
+        let mut batch: ArrayVec<(Ethernet2Header, DemiBuffer), RECEIVE_BATCH_SIZE> = ArrayVec::new();
+        for pkt in self.layer1_endpoint.receive()? {
+            let (header, payload) = match Ethernet2Header::parse(pkt) {
+                Ok(result) => result,
+                Err(e) => {
+                    // TODO: Collect dropped packet statistics.
+                    let cause: &str = "Invalid Ethernet header";
+                    warn!("{}: {:?}", cause, e);
+                    continue;
+                },
+            };
+            debug!("Engine received {:?}", header);
+            if self.link_addr != header.dst_addr()
+                && !header.dst_addr().is_broadcast()
+                && !header.dst_addr().is_multicast()
+            {
+                let cause: &str = "invalid link address";
+                warn!("dropping packet: {}", cause);
+            }
+            batch.push((header, payload))
+        }
+        Ok(batch)
+    }
+
+    pub fn transmit<P: PacketBuf>(&mut self, mut pkt: P) -> Result<(), Fail> {
+        let buf: DemiBuffer = match pkt.take_body() {
+            Some(body) => body,
+            _ => {
+                let cause = format!("No body in PacketBuf to transmit");
+                warn!("{}", cause);
+                return Err(Fail::new(libc::EINVAL, &cause));
+            },
+        };
+        self.layer1_endpoint.transmit(buf)
+    }
+}
+
+//======================================================================================================================
+// Trait Implementations
+//======================================================================================================================
+
+impl Deref for SharedLayer2Endpoint {
+    type Target = Layer2Endpoint;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.deref()
+    }
+}
+
+impl DerefMut for SharedLayer2Endpoint {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0.deref_mut()
+    }
+}
+
+/// Memory Runtime Trait Implementation for the network stack.
+impl MemoryRuntime for Layer2Endpoint {
+    /// Casts a [DPDKBuf] into an [demi_sgarray_t].
```

Comment:
These (and such) comments can be skipped because they are not adding any more context than the code below (a few exceptions are: externally facing APIs, very complex pieces of code etc.). However, in this case, this function delegates the responsibility of casting further and doesn't do it here itself. These comments are susceptible to deviation from the code in future. I understand that we want to produce good rust docs, but we can be selective about that, so whenever we have any comments, they add a meaningful context to be consumed in conjunction with reading the code - otherwise, code is the best documentation. Also, I doubt that anybody would just read rust docs/comments without the need of reading the code in order to understand how the system works.